charmcraft/commands/pack.py

```diff
@@ -0,0 +1,130 @@
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+import logging
+import os
+import pathlib
+import zipfile
+
+from charmcraft.cmdbase import BaseCommand, CommandError
+from .utils import load_yaml
+
+logger = logging.getLogger(__name__)
+
+# the minimum set of files in a bundle
+MANDATORY_FILES = {'bundle.yaml'}
+
+
+def build_zip(zippath, basedir, fpaths):
+    """Build the final file.
+
+    Note we convert all paths to str to support Python 3.5.
+    """
+    zipfh = zipfile.ZipFile(str(zippath), 'w', zipfile.ZIP_DEFLATED)
+    for fpath in fpaths:
+        zipfh.write(str(fpath), str(fpath.relative_to(basedir)))
+    zipfh.close()
+
+
+def get_paths_to_include(dirpath):
+    """Get all file/dir paths to include."""
+    allpaths = set()
+
+    # all mandatory files, which must exist (currently only bundles.yaml is mandatory, and
+    # it's verified before)
+    for fname in MANDATORY_FILES:
+        allpaths.add(dirpath / fname)
+
+    # the extra files, which must be relative
+    config = load_yaml(dirpath / 'charmcraft.yaml') or {}
+    prime_specs = config.get('parts', {}).get('bundle', {}).get('prime', [])
+
+    for spec in prime_specs:
+        if spec[0] == os.path.sep:
+            raise CommandError(
+                "Extra files in prime config can not be absolute: {!r}".format(spec))
+
+        fpaths = sorted(fpath for fpath in dirpath.glob(spec) if fpath.is_file())
+        logger.debug("Including per prime config %r: %s.", spec, fpaths)
+        allpaths.update(fpaths)
+
+    return sorted(allpaths)
+
+
+_overview = """
+Build the bundle, leaving a .zip file as the result of the process.
+
+You can `juju deploy` directly from the resulting .zip file, or upload it to
+the store (see the "upload" command).
+"""
+
+
+class PackCommand(BaseCommand):
+    """Build the bundle or the charm.
+
+    Eventually this command will also support charms, but for now it will work only
+    on bundles.
+    """
+    name = 'pack'
+    help_msg = "Build the bundle"
+    overview = _overview
+
+    def fill_parser(self, parser):
+        """Add own parameters to the general parser."""
+        parser.add_argument(
+            '-f', '--from', type=pathlib.Path, dest='from_dir',
+            help="The directory where the bundle project is located, where the build "
+                 "is done from; defaults to '.'")
+
+    def run(self, parsed_args):
+        """Run the command."""
+        if parsed_args.from_dir is None:
+            dirpath = pathlib.Path.cwd()
+        else:
+            dirpath = parsed_args.from_dir.expanduser()
+            if not dirpath.exists():
+                raise CommandError("Bundle project directory was not found: '{}'.".format(dirpath))
+            if not dirpath.is_dir():
+                raise CommandError(
+                    "Bundle project directory is not really a directory: '{}'.".format(dirpath))
```

Comment:
```suggestion
                    "Bundle project directory is not a directory: {!r}.".format(str(dirpath)))
```

Reply:
Removed the "really". Regarding how to show dirpath, this is for consistency with the rest of the project.

I'm +1 to revisit that decision (if you want) and propose a branch to change them all (if needed).

Reply:
Yeah, that's the right call. I thought about that afterwards and forgot to come back.  `!r` is the approach made throughout most of snapcraft.

Reply:
Opened https://github.com/canonical/charmcraft/issues/208 for this